docs/rfcs/templates_solutions.md

```diff
@@ -0,0 +1,314 @@
+    CHOP_PCK
+
+    PackageVersions
+    -----------------
+    CHOP_PCK.package_versions = [CHOP_PCK.v1, CHOP_PCK.v2]
+
+    CHOP_PCK.v1.template_versions = [PDE.v1, BCM.v1]
+    CHOP_PCK.v2.template_versions = [PDE.v1, BCM.v2]
+    ```
+
+7. Assign new package version to CBTN
+
+    ```
+    Study-CBTN.package_versions = [CHOP_PCK.v2]
+    Study-A.package_versions = [CHOP_PCK.v1]
+    ```
+
+### Template Management Rules
+
+1. A Template / TemplateVersion may be created at any time
+2. A Template / TemplateVersion may be updated/deleted if it is not being used in any study
+3. A TemplatePackage / PackageVersion may be created at any time
+4. A PackageVersion may be assigned to a study at any time
+5. TemplateVersions may be added/updated/deleted in a TemplatePackage via
+   creating a new PackageVersion with the desired TemplateVersions
+
+### Permissions
+
+1. Any organization or study admin may create Template/TemplateVersions
+2. Any organization or study admin may add TemplateVersions to a Package owned by their organization
+3. Any organization or study admin may update/delete TemplateVersions owned by their organization
+4. Any organization or study admin may assign PackageVersions to a study they have access to
+5. Any organization or study admin may update a TemplatePackage owned by their organization
+
+### Package Versioning (Semantic Versions)
+
+Automatically generatig a semantic version number for the template package
+version is difficult. Automatic semantic versioning requires an agreed upon
+set of rules by all organizations that are creating template packages.
+Even if the example rules below were used to generate semantic version numbers,
+the backend needs to understand package updates at a deep level of granularity
+in order to generate meaningful semantic version numbers.
+
+We would most likely want to use these rules as guidelines or standards and
+allow users to update the semantic version number of their package.
+
+#### Major Version Updates
+
+1. Add a new template (new Template) to the template package to capture a new type of data
+1. Delete a template from the template package
+
+#### Minor Version Updates
+
+1. Add a new template version (new TemplateVersion) with field schema updates to
+   an existing template within the package
+
+#### Patch Version Updates
+
+1. Add a new template version (new TemplateVersion) with basic info updates
+   to an existing template within the package
+
+### Pros/Cons
+
+-   ✅ Templates are easily shareable among studies
+-   ✅ Can reliably/programmatically track provenance of templates
+-   ✅ Studies are not affected by non-relevant template updates
+-   ✅ Easier to manage a set of logically related templates and add to multiple studies
+-   ✅ Easier to track / capture a batch of template updates using template package versions
+-   ⚠️ UI design and frontend implementation becomes considerably more complex
+-   ⚠️ Backend implementation becomes considerably more complex
```

Comment:
Recall that the requirement for schema enforcement on uploaded files comes from the need to be able to shovel this data into downstream pipelines. Considering that, it seems like each organization will naturally have one or two 'base' models that each study will need to adhere to. This seems to work pretty well with the package concept. However, there is a further need to have the supplemental fields, perhaps per-study. So each study will match up with the latest (hopefully) package version, but may have a handful of 'supplemental' templates that are specific to that study or the requirements of that study.

Perhaps we could consider these supplemental templates for the future. Going this way would probably require one package per study with the goal being that all studies really only target the latest version of that package. If we go with package versions, maybe we should consider restricting each study to only implement one package at a time.

Reply:
Yea I agree - a Study has one PackageVersion (aka foreign key) at a time. I thought that was represented in the model diagram. If not, that's a mistake.